### PR TITLE
Updated to react 18 and also updated other libraries connected

### PR DESCRIPTION
**Pull request by hkurhinen**

### CODE REVIEW

**Review by EVP1989 — COMMENTED** [review]
Added comments

File: src/components/SeedBatchList.tsx [hunk 1]
```diff
@@ -16,13 +16,13 @@ import {
...
 } from "semantic-ui-react";
-import LocalizedUtils from "src/localization/localizedutils";
+import LocalizedUtils from "../localization/localizedutils";
 
 export interface Props {
   keycloak?: Keycloak.KeycloakInstance;
   seedBatches?: SeedBatch[];
   onSeedBatchesFound?: (seedBatches: SeedBatch[]) => void,
-  onError: (error: ErrorMessage) => void
+   onError: (error: ErrorMessage | undefined) => void
```
Comment (Jdallos) [inline src/components/SeedBatchList.tsx:25]:
```suggestion
  onError: (error: ErrorMessage | undefined) => void
```

File: src/components/SeedBatchList.tsx [hunk 2]
```diff
@@ -182,7 +182,7 @@ export function mapStateToProps(state: StoreState) {
 export function mapDispatchToProps(dispatch: Dispatch<actions.AppAction>) {
   return {
     onSeedBatchesFound: (seedBatches: SeedBatch[]) => dispatch(actions.seedBatchesFound(seedBatches)),
-    onError: (error: ErrorMessage) => dispatch(actions.onErrorOccurred(error))
+     onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error))
```
Comment (Jdallos) [inline src/components/SeedBatchList.tsx:185]:
```suggestion
    onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error))
```

File: src/components/SeedList.tsx [hunk 3]
```diff
@@ -15,13 +15,13 @@ import {
...
 } from "semantic-ui-react";
-import LocalizedUtils from "src/localization/localizedutils";
+import LocalizedUtils from "../localization/localizedutils";
 
 export interface Props {
   keycloak?: Keycloak.KeycloakInstance;
   seeds?: Seed[];
   onSeedsFound?: (seeds: Seed[]) => void,
-  onError: (error: ErrorMessage) => void
+   onError: (error: ErrorMessage | undefined) => void
```
Comment (Jdallos) [inline src/components/SeedList.tsx:24]:
```suggestion
  onError: (error: ErrorMessage | undefined) => void
```

File: src/components/SeedList.tsx [hunk 4]
```diff
@@ -132,7 +132,7 @@ export function mapStateToProps(state: StoreState) {
 export function mapDispatchToProps(dispatch: Dispatch<actions.AppAction>) {
   return {
     onSeedsFound: (seeds: Seed[]) => dispatch(actions.seedsFound(seeds)),
-    onError: (error: ErrorMessage) => dispatch(actions.onErrorOccurred(error)),
+     onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error)),
```
Comment (Jdallos) [inline src/components/SeedList.tsx:135]:
```suggestion
    onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error)),
```

File: src/components/WastageReasonList.tsx [hunk 5]
```diff
@@ -24,7 +24,7 @@ interface Props {
   keycloak?: Keycloak.KeycloakInstance;
   wastageReasons?: WastageReason[];
   onWastageReasonsFound?: (wastageReasons: WastageReason[]) => void,
-  onError: (error: ErrorMessage) => void
+   onError: (error: ErrorMessage | undefined) => void
```
Comment (Jdallos) [inline src/components/WastageReasonList.tsx:27]:
```suggestion
  onError: (error: ErrorMessage | undefined) => void
```

File: src/components/WastageReasonList.tsx [hunk 6]
```diff
@@ -144,7 +144,7 @@ export function mapStateToProps(state: StoreState) {
 export function mapDispatchToProps(dispatch: Dispatch<actions.AppAction>) {
   return {
     onWastageReasonsFound: (wastageReasons: WastageReason[]) => dispatch(actions.wastageReasonsFound(wastageReasons)),
-    onError: (error: ErrorMessage) => dispatch(actions.onErrorOccurred(error))
+     onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error))
```
Comment (Jdallos) [inline src/components/WastageReasonList.tsx:147]:
```suggestion
    onError: (error: ErrorMessage | undefined) => dispatch(actions.onErrorOccurred(error))
```

File: src/components/WelcomePage.tsx [hunk 7]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/packings/:packingId"
-                exact={true}
-                render={props => (
-                  <EditPacking
-                    keycloak={this.state.keycloak}
-                    packingId={props.match.params.packingId as string}
-                />
-                )}
+                element={
+                  <WithParams Component={EditPacking} keycloak={this.state.keycloak} routeParamNames={["packingId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:233]:
```suggestion
                  <WithParams Component={ EditPacking } keycloak={ this.state.keycloak } routeParamNames={ ["packingId"] } />
```

File: src/components/WelcomePage.tsx [hunk 8]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/cutPackings/:cutPackingId"
-                exact={ true }
-                render={ props => (
-                  <EditCutPacking
-                    keycloak={ this.state.keycloak }
-                    cutPackingId={ props.match.params.cutPackingId as string }
-                  />
-                ) }
+                element={
+                  <WithParams Component={EditCutPacking} keycloak={this.state.keycloak} routeParamNames={["cutPackingId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:239]:
```suggestion
                  <WithParams Component={ EditCutPacking } keycloak={ this.state.keycloak } routeParamNames={ ["cutPackingId"] } />
```

File: src/components/WelcomePage.tsx [hunk 9]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/campaigns/:campaignId"
-                exact={ true }
-                render={ props => (
-                  <EditCampaign
-                    keycloak={ this.state.keycloak }
-                    campaignId={ props.match.params.campaignId as string }
-                />
-                )}
+                element={
+                  <WithParams Component={EditCampaign} keycloak={this.state.keycloak} routeParamNames={["campaignId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:245]:
```suggestion
                  <WithParams Component={ EditCampaign } keycloak={ this.state.keycloak } routeParamNames={ ["campaignId"] } />
```

File: src/components/WelcomePage.tsx [hunk 10]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/discards/:discardId"
-                exact
-                render={ props => (
-                  <EditDiscard
-                    keycloak={ this.state.keycloak }
-                    discardId={ props.match.params.discardId as string }
-                  />
-                )}
+                element={
+                  <WithParams Component={EditDiscard} keycloak={this.state.keycloak} routeParamNames={["discardId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:251]:
```suggestion
                  <WithParams Component={ EditDiscard } keycloak={ this.state.keycloak } routeParamNames={ ["discardId"] } />
```

File: src/components/WelcomePage.tsx [hunk 11]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/products/:productId"
-                exact={true}
-                render={props => (
-                    <EditProduct
-                      keycloak={this.state.keycloak}
-                      productId={props.match.params.productId as string}
-                    />
-                )}
+                element={
+                  <WithParams Component={EditProduct} keycloak={this.state.keycloak} routeParamNames={["productId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:289]:
```suggestion
                  <WithParams Component={ EditProduct } keycloak={ this.state.keycloak } routeParamNames={ ["productId"] } />
```

File: src/components/WelcomePage.tsx [hunk 12]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/createEvent"
-                exact={true}
-                render={props => (
-                    <CreateEvent
-                      keycloak={this.state.keycloak}
-                    />
-                )}
+                element={
+                  <CreateEvent
+                    keycloak={this.state.keycloak}
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:302]:
```suggestion
                    keycloak={ this.state.keycloak }
```

File: src/components/WelcomePage.tsx [hunk 13]
```diff
@@ -171,387 +172,308 @@ class WelcomePage extends React.Component<Props, any> {
...
                 path="/pests/:pestId"
-                exact={true}
-                render={props => (
-                    <EditPest
-                      keycloak={this.state.keycloak}
-                      pestId={props.match.params.pestId as string}
-                    />
-                )}
+                element={
+                  <WithParams Component={EditPest} keycloak={this.state.keycloak} routeParamNames={["pestId"]} />
```
Comment (Jdallos) [inline src/components/WelcomePage.tsx:369]:
```suggestion
                  <WithParams Component={ EditPest } keycloak={ this.state.keycloak } routeParamNames={ ["pestId"] } />
```

File: src/components/WithParams.tsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  Component,
+  keycloak,
+  routeParamNames
+}) => {
+  const params = useParams();
+  let paramMap: any = {}
+  routeParamNames.forEach(name => {
+    paramMap[name] = params[name];
+  });
+  return <Component keycloak={keycloak} {...paramMap} />
```
Comment (Jdallos) [inline src/components/WithParams.tsx:20]:
```suggestion
  return <Component keycloak={ keycloak } { ...paramMap } />
```

**Review by Jdallos — APPROVED** [review]
Only suggestions are minor spacing changes.

**Pull request merged**